### DISCUSSION TITLE
Is it possible to avoid public hoisting (`nohoist` behaviour from yarn workspaces) with `node-linker=hoisted`?

### DISCUSSION THREAD

**Discussion by rrebase** [pnpm/pnpm#5651] [Q&A]
In React Native Monorepo setups setting `node-linker=hoisted` is a must to have module resolution play well with metro. However, seems that there's no way in pnpm to avoid public hoisting (deps being installed in root node_modules instead of project) similar to what yarn workspaces `nohoist` would do, is that true? Anyway to disable it?

The following does not have any effect with `node-linker=hoisted`
```
public-hoist-pattern[]='!@types/react'
hoist-pattern[]='!*eslint-plugin*'
```

**Comment by rrebase** [comment]
@zkochan do you have any thoughts/input here? 🙏

**Reply by zkochan** [reply]
Do you mean the hoisting limits setting of Yarn? 
https://yarnpkg.com/configuration/manifest#installConfig.hoistingLimits

**Reply by rrebase** [reply]
I meant that none of the pnpm config options to **not** hoist packages by pattern won't work if `node-linker` is set to `hoisted`.

But yes, based on the description, it would also work if pnpm supported an equivalent to the Yarn `hoistingLimits` 🙂

**Reply by markh123** [reply]
I am running into this as well. What I need is the ability to not hoist a specific package when `node-linker=hoisted`.  I am not sure the `yarn` `hoistingLimits` is ideal for my use case considering it applies to the entire package.  I just need to prevent certain packages from being hoisted.  If we could make negative `hoist-pattern` or negative `public-hoist-pattern` work when `node-linker=hoisted` that would be ideal.   I guess this is similar to the `yarn v1` [nohoist](https://classic.yarnpkg.com/blog/2018/02/15/nohoist/).

**Reply by rrebase** [reply]
@zkochan Sorry to 🏓 again, but have you had the chance to investigate this?

**Reply by zkochan** [reply]
`public-hoist-pattern` will not work with `node-linker=hoisted`

For hoisting limit here is a PR: https://github.com/pnpm/pnpm/pull/6468

**Reply by merojosa** [reply]
@rrebase, did you use a workaround for this situation?

**Reply by rrebase** [reply]
@merojosa yep, used temporary workarounds depending on the dep that blew up. In one case locked same package to different patch versions so that it wouldn't be hoisted 🙈 

Haven't tried myself yet. But with the latest metro version supporting [unstable_enableSymlinks](https://reactnative.dev/blog/2023/06/21/0.72-metro-package-exports-symlinks#symlink-support-beta) might be able to drop `node-linker=hoisted` altogether

**Reply by 0xPT** [reply]
@rrebase this worked for me! I have a React Native/Expo app and another app that I need to nohoist in the same monorepo. The new metro feature allowed me to remove node-linker=hoisted.

**Comment by spigelli** [comment]
https://pnpm.io/npmrc#hoist-pattern

**Reply by rrebase** [reply]
Unfortunately this does not work as described if using `node-linker=hoisted`

**Comment by hydra1983** [comment]
I have to set `shared-workspace-lockfile=false` in `.npmrc` to fix this problem
